hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientClusterServiceImpl.java

```diff
@@ -288,35 +235,38 @@ private MemberListSnapshot createSnapshot(int memberListVersion, Collection<Memb
                     .memberListJoinVersion(memberInfo.getMemberListJoinVersion());
             newMembers.put(memberInfo.getUuid(), memberBuilder.build());
         }
-        return new MemberListSnapshot(memberListVersion, newMembers);
+        return new MemberListSnapshot(memberListVersion, newMembers, clusterUuid);
     }
 
     private Set<Member> toUnmodifiableHasSet(Collection<Member> members) {
         return unmodifiableSet(new HashSet<>(members));
     }
 
-    private List<MembershipEvent> detectMembershipEvents(Collection<Member> prevMembers, Set<Member> currentMembers) {
+    private List<MembershipEvent> detectMembershipEvents(Collection<Member> prevMembers,
+                                                         Set<Member> currentMembers,
+                                                         UUID clusterUuid) {
         List<Member> newMembers = new LinkedList<>();
         Set<Member> deadMembers = new HashSet<>(prevMembers);
-        for (Member member : currentMembers) {
-            if (!deadMembers.remove(member)) {
-                newMembers.add(member);
+
+        if (clusterUuid.equals(memberListSnapshot.get().clusterUuid)) {
+            for (Member member : currentMembers) {
+                if (!deadMembers.remove(member)) {
+                    newMembers.add(member);
+                }
             }
+        } else {
+            // if cluster uuid is not same, then we will not try to match the current members to previous members
+            // As a result all previous members are dead and all current members are new members
```

Comment:
Is there a behaviour change here? Will the user receive member removed event for the member in the old cluster? If so, how?

Reply:
The behaviour change is as descriped in the pr description.
 ```
Now clusterUuid is part of the logic. We don't clear the memberlist at all. Upon receiving,
an event, we compare existing and new clusterUuid. If they are different,
without comparing the members, we fire removed events for all existing
members, and added events for all new ones to achieve same behaviour.
```